scripts/versions-replacer/src/replace.rs

```diff
@@ -0,0 +1,88 @@
+use std::{borrow::Cow, collections::HashMap, fs, path::Path};
+
+use color_eyre::{eyre::Context, Result};
+use once_cell::sync::Lazy;
+use regex::{Captures, Regex};
+
+pub static VERSIONS_REGEX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"\{\{versions\.([\w_-]+)\}\}").unwrap());
+
+pub fn replace_versions_in_file(
+    path: impl AsRef<Path>,
+    versions: &HashMap<String, String>,
+) -> Result<Option<usize>> {
+    let path = path.as_ref();
+    let contents =
+        fs::read_to_string(path).wrap_err_with(|| format!("failed to read {:?}", path))?;
+    let (replaced_contents, replacement_count) = replace_versions_in_string(&contents, versions);
+    if replacement_count.is_some() {
+        fs::write(path, replaced_contents.as_bytes())
+            .wrap_err_with(|| format!("failed to write back to {:?}", path))?;
+    }
+    Ok(replacement_count)
+}
+
+pub fn replace_versions_in_string<'a>(
+    s: &'a str,
+    versions: &HashMap<String, String>,
+) -> (Cow<'a, str>, Option<usize>) {
```

Comment:
Why not signal 'no changes' with `0usize`? 

Write the file only if `num_changes != 0`.